auto_post_classifier/gpt_handler.py

```diff
@@ -0,0 +1,199 @@
+import json
+import jsonschema
+from api_request_parallel_processor import process_api_requests
+from loguru import logger
+from pathlib import Path
+
+import consts
+
+class GPT_MODEL:
+    GPT_3_5_16k = "gpt-3.5-turbo-16k"
+    GPT_4 = ""
+
+class ResponseValidator:
+
+    def __init__(self) -> None:
+        self.validators = [
+            self.validate_json,
+            self.validate_explenation
+        ]
+        
+
+    def validate_json(self, response_dict: dict):
+    # Define the JSON schema
+        schema = {
+            "type": "object",
+            "properties": {"summary": {"type": "string"}},
+            "patternProperties": {
+                ".*_exp$": {"type": "string"},
+                ".*_rnk$": {"type": "number"},
+            },
+            "additionalProperties": False,
+            "required": ["summary"],
+        }
+
+        # Validate the JSON data against the schema
+        try:
+            jsonschema.validate(instance=response_dict, schema=schema)
+            return True
+        except jsonschema.exceptions.ValidationError as e:
+            logger.warning(f"JSON is not valid: {e.message}")
+            return False
+
+    def validate_explenation(self, response_dict: dict):
+        for dimension in consts.WEIGHTS:
+            if response_dict[f"{dimension}_exp"] == f"-1, 0, or 1 ranking for {dimension}":
+                logger.warning("validation error")
+                return False
+        return True        
+
+    def validate(self, response: str)  -> (bool, str, dict) :
+        try:
+            response_dict = json.loads(response)
+            for validator in self.validators:
+                if not validator(response_dict):
+                    return (False, validator.__name__, response_dict)
+
+                return (True, "", response_dict)
+            
+        except Exception as e:
+            logger.exception("validation exception:", e)
+class GptHandler:
+
+    def build_prompt(self, text):
+        user_prompt = f"""Here is some context:
+@@In recent news, a bold and unprecedented attack was carried out by the Palestinian militant group Hamas against Israel.
+This egregious assault involved approximately 3,000 assailants infiltrating Israeli territory,
+resulting in numerous casualties, including both soldiers and civilians. Dozens of hostages were also abducted.
+The attack has been marked by extreme brutality, with disturbing reports of massacres, beheadings, sexual violence,
+and the use of civilians as human shields. These actions have been widely condemned as war crimes and crimes against
+humanity under international humanitarian law.@@
+
+The result should be provided as a JSON with the following format:
+
+{{
+    "antisemitism_exp": "Explanation regarding the presence of antisemitic content",
+    "antisemitism_rnk": "-1, 0, or 1 ranking for antisemitism",
```

Comment:
I think we will right the tests for the promts and then we will try to test some changes in the prompt. what do you think?

Reply:
I don't have experience with this, but my guess is that it's not straightforward to "test" the responses of a GPT model, because it is not deterministic.

Reply:
i am talking about the tests of the validation of the response first. I hope I will manage to do it today. the general idea is to take about 10  posts and then run them and check what validation went wrong. with this, we can compare the explanation in the sys prompt or in user prompt